gradio/external.py

```diff
@@ -417,6 +419,35 @@ def get_spaces(model_name, api_key, alias, **kwargs):
         return get_spaces_blocks(model_name, config)
 
 
+async def get_pred_from_ws(
```

Comment:
Split this out into it's own function as opposed to defining it in `get_ws_fn` so that it's easier to test. It's tricky mocking async context generators and I sunk way too much time into it.